### ISSUE TITLE
Adaptive Gradient Clipping

### ISSUE THREAD

**Issue by benihime91** [lessw2020/Ranger21#10]
Hi @lessw2020 thanks for this awesome work . I came here from the fastai forums and have been playing around with `Ranger21` for a few days now. The results seem pretty solid and in most cases I was easily able to beat `Ranger` or get comparable results. Just a few points I noticed ...

1. I don't think AGC is working if we train using `fp16`. I was getting some weird losses if I kept `use_adaptive_gradient_clipping` on while training in `fp16`. I works fine if I keep training in `fp32` though. Is this something to be expected or am I doing something wrong ?
2.  I also noticed that the `learning rate` of paramters in Ranger21 is not modified i.e., `optimizer.param_groups[n]["lr"]` remains same throughout. Are you computing the learning rate schedule on the fly and then updating the weights ?

**Comment by lessw2020** [comment]
Hi @benihime91,
Thanks for the feedback and great to have you here.
Glad to hear you are seeing outperformance with Ranger21 vs original Ranger, since that's the goal of Ranger21 of course.
To your questions:
1 - Would not expect agc to do well in fp16.  Thus, would not recommend using it during fp16 training.  
That said, in theory the optimization step including agc, should all still be happening in fp32.  I am no expert on this topic, but I thought that only specific steps are done in fp16 (i.e. forward and backward) but that the optimization step would still happen in fp16.  Can I ask what you are using to run fp16 (i.e. apex or ?).

2 - Re: lr - yes atm the lr is adjusted dynamically on each step rather than being stored at a fixed point.  The idea is that the only key thing we need is the base lr, and then the adjustments are all computed on the fly based on your settings for Ranger21 and where it is in the total training lifecycle.

**Comment by benihime91** [comment]
Thanks for the reply @lessw2020 . 
1. It was actually a bug in my `fp16` training code. Big apologies. What you said made sense so I went back and checked my code and there it was. 
2. Also why would would you think `Would not expect agc to do well in fp16` ? Any insights, if possible ?  

Thanks. In the meantime I am closing this issue. ☺️

**Comment by lessw2020** [comment]
Hi @benihime91  - the main issue with fp16 and agc would be that we are using the gradient norm to parameter norm to calculate it.  If you use fp16, then the gradient norm would be much smaller and so at a minimum, the hyperparameter for it would need to be re-tuned every dataset.
In addition, normally for fp16 you do dynamic gradient scaling, meaning the scaling multiple is being adjusted iteration to iteration in order to maximize the range that the fp16 covers (i.e. as wide as possible, but without overflow).  
However, it seems that would make a moving target for the gradient clipping with the constant rescaling, and that would be the primary issue imo.  
That said, very glad to hear that the core issue was ensuring that the optimization step was done in fp32 - that should let you get the best of both.  
Thanks for the update on this issue and hope you are doing well!
